src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1797,15 +1797,36 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             $container->setAlias('messenger.default_serializer', $config['serializer']['default_serializer']);
         }
 
+        $failureTransports = [];
+        if ($config['failure_transport']) {
+            if (!isset($config['transports'][$config['failure_transport']])) {
+                throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $config['failure_transport']));
+            }
+
+            $container->setAlias('messenger.failure_transports.default', $config['failure_transport']);
+            $failureTransports[] = $config['failure_transport'];
+        }
+        foreach ($config['transports'] as $name => $transport) {
+            if (isset($transport['failure_transport']) && !isset($config['transports'][$transport['failure_transport']])) {
+                throw new LogicException(sprintf('Invalid Messenger configuration: the failure transport "%s" is not a valid transport or service id.', $transport['failure_transport']));
+            }
+            if ($transport['failure_transport']) {
+                $failureTransports[] = $transport['failure_transport'];
```

Comment:
What if we create a 2nd array right here?

```
if ($transport['failure_transport']) {
    $failureTransports[] = $transport['failure_transport'];
    $failureTransportsByName[$name] = $transport['failure_transport'];
} elseif ($config['failure_transport']) {
    // no need to add it to $failureTransports, as we will always set it above
    $failureTransportsByName[$name] = $config['failure_transport'];
}
```

The advantage is that, below, around 1860, instead of looping again over `$config['transports']`, we could loop over `$failureTransportsByName[$name]`. This would allow us to remove most of the logic from that foreach, except for the part that uses the `$senderReferences`. And the `LogicException` down there would check the validity of *all* of the failure transports, including the global one (because as long as the global transport should be used for 1 transport, it will already be in `$failureTransportsByName`.

Heck, the `foreach` around 1860 could probably just be an `array_map` :p (notice I renamed the `$failureTransportsByName` variable from around 1860 to `$failureTransportReferencesByTransportName`. Also, that logic could now possibly go down around line 1913 - it's only needed for the `ServiceLocatorTagPass::register` call.

```
$failureTransportReferencesByTransportName = array_map(function () {...}, $failureTransportsByName);
```